Gcp projects (#148)
nais/doc: docs/basics/teams.md

### Rotate API key

Go to [deploy.nais.io](https://deploy.nais.io/) and click on "Create new key" button for

## Team namespaces

Team namespaces are supported in both on-prem and in GCP. Refer to the [team namespaces documentation](../clusters/team-namespaces.md) for details.

## GCP Team projects

Each team has their own [Google Cloud Platform project](https://cloud.google.com/storage/docs/projects), which consist of a set of users; a set of APIs; and authentication, and monitoring settings of those APIs.
So, for example, all of your Cloud Storage buckets and objects, along with user permissions for accessing them, reside in a project.

In general every memeber of the team have the possibility to add the necessary permissions they need via [IAM role managment](https://console.cloud.google.com/iam-admin/iam).
There are also no general limit on which features/products that can be used in a project, but everything has need been [ROS'd](https://doc.nais.io/legal/nais-ros/).
Each team is responsible to do the necessary ROS and PVK changes to use a new feature.

> Remember to clean up after yourself, so we don't have unnecessary resources running, and costing us money.

Author: Kunne vi funnet et annet ord enn `leveraged` her? Måtte slå det opp, men hjalp meg ikke i å forstå setningen din.

> Remember to clean up after yourself, so we don't have unnecessary resources running, and costing us money.

Reviewer: Tror setningen i forslaget mitt fungerer uten ordet leveraged, når jeg tenker meg om...

We have a dashboard showing what each [team](https://datastudio.google.com/u/1/reporting/417b0a1d-b307-4a6d-a699-77a6ab239661/page/mJdmB) is using, plus a dashboard for [everything](https://datastudio.google.com/reporting/fda5f821-caef-4056-9356-9aa4f7082699/page/mJdmB) in GCP.

### Access management

To promote autonomous teams, teams control access to their own projects completely by them self.
This can be done either by pipeline, or manually adding access when needed.

There are different scenarios for when and how to give access to users, and the official [Google Cloud Docs](https://cloud.google.com/iam/docs/granting-changing-revoking-access) is definitive the best source for information.

> Google Docs has a list of possible [predefined roles](https://cloud.google.com/iam/docs/understanding-roles#predefined_roles) that we recommend using.

Author: Kanskje forenkle `keeping in mind/teams leverage` til å ikke ha `/`, blir litt enklere å lese da. For eksempel bare `using`. Da jeg vil tro det beste er at folk gjenbruker de rollene, i stedet for å prøve å lage sine egne.

> Google Docs has a list of possible [predefined roles](https://cloud.google.com/iam/docs/understanding-roles#predefined_roles) that we recommend using.

Reviewer: `keep in mind/use.` burde funke? Hva tenker du? =)

> Google Docs has a list of possible [predefined roles](https://cloud.google.com/iam/docs/understanding-roles#predefined_roles) that we recommend using.

Author: Funker bedre, slipper å lese hele setningen på nytt.

> These roles can also be listed our with `gcloud iam roles list --filter $feature`.

Author: Kan prøve å lage en mer beskrivende tekst. Filteret var ikke så veldig bra. Kan hende den enkle løsningen er å bytte det `feature` med `roleName`.

Feature er vel uansett ikke forklart i denne konteksten, som er litt dumt. Kan hende, på et generelt basis, bør bytte det ut med `product`.

Running the command without the `--filter` argument will return a very long list.

#### Examples

!!! Info "Access to Postgres"
    We have a separate guide for how to give your self temporary access to a Postgres database. Head on over to [persistence/postgres](../persistence/postgres/#granting-temporary-personal-access).

Underneath we have outlined two typical use cases for when you need to give your self or someone else access.
Do get a better understanding of how to give and revoke access with `gcloud`, you should read [Granting, changing, and revoking access to resources
](https://cloud.google.com/iam/docs/granting-changing-revoking-access).

##### How to give your self access

To be able to run this commando, you first need to find your `PROJECT_ID`, and the specific `ROLE_NAME` you want access to.
```bash

> gcloud projects add-iam-policy-binding <PROJECT_ID> --member=user:<FIRSTNAME>.<LASTNAME>@nav.no --role=<ROLE_NAME> --condition="expression=request.time < timestamp('$(date -v '+1H' -u +'%Y-%m-%dT%H:%M:%SZ')'),title=temp_access"

Author: Tenker vi flytte bash-kommentaren din ut i teksten ovenfor, så det ikke blir så mye gjentakelse.

> gcloud projects add-iam-policy-binding <PROJECT_ID> --member=user:<FIRSTNAME>.<LASTNAME>@nav.no --role=<ROLE_NAME> --condition="expression=request.time < timestamp('$(date -v '+1H' -u +'%Y-%m-%dT%H:%M:%SZ')'),title=temp_access"

Reviewer: Tar du den du? Får ikke enkelt editert mer enn én linje sånn som dette fungerer...
Bryr meg lite om det er utenfor eller inni (kommentaren).

```

For example if you wanted to take a look at the Storage buckets that you have, then you would need the `roles/storage.objectViewer` role.
It's always smart to only give out [temporary access](#temporary-access).

This can also be used to give a user in a different team access.

##### How to give a service account access

> A service account is also a `--member` of a project, but instead of running the command using in the previous paragraph, you change `--member=user:<email>` to `--member=serviceAccount:<email>`.

Author: Hva ønsker du å teste lokalt?

> A service account is also a `--member` of a project, but instead of running the command using in the previous paragraph, you change `--member=user:<email>` to `--member=serviceAccount:<email>`.

Reviewer: At lenken til markdown header innad samme fil fungerer translert til http

> #### Temporary access

Author: God idé!

On a general basis, access should not be permanent.
A good habit is to only grant your self or other a temporary access.

Using `gcloud`-cli the following command will grant a user 1 hours of access to `roles/cloudsql.instanceUser`:
```bash
gcloud projects add-iam-policy-binding <PROJECT_ID> --member=user:<FIRSTNAME>.<LASTNAME>@nav.no --role=roles/cloudsql.instanceUser --condition="expression=request.time < timestamp('$(date -v '+1H' -u +'%Y-%m-%dT%H:%M:%SZ')'),title=temp_access"
```

There is more information over at [Google Cloud Docs](https://cloud.google.com/iam/docs/configuring-temporary-access).

#### IAM recommander

> Some time you end up giving a wider access than needed, but GCP has a IAM recommender that will monitor each access given, and compare it to the access used.

Reviewer: ```suggestion
Sometimes you might end up giving more (a wider) access than strictly - but fear not!
GCP has an IAM recommender which will monitor each access permission given, and compare it to accesses actually asked for/required aftwards.
```

Author: Foreslår å droppe sammenslåingen ved bruk av `/`, syns teksten blir tungvindt å lese da.

> Then the IAM recommender will recommend a more granular access, that is probably more fitting for the user.

Reviewer: ```suggestion
After which the IAM recommender will recommend a more granular access, that probably fits you and your usage better!
```
